include/client.h

```diff
@@ -0,0 +1,116 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MIXERCLIENT_CLIENT_H
+#define MIXERCLIENT_CLIENT_H
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "google/protobuf/stubs/status.h"
+#include "mixer/api/v1/service.pb.h"
+#include "options.h"
+
+namespace google {
+namespace mixer_client {
+
+// Defines a function prototype used when an asynchronous transport call
+// is completed.
+using DoneFunc = std::function<void(const ::google::protobuf::util::Status&)>;
+
+// Defines a function prototype to make an asynchronous Check call to
+// the mixer server.
+using TransportCheckFunc = std::function<void(
+    const ::istio::mixer::v1::CheckRequest& request,
+    ::istio::mixer::v1::CheckResponse* response, DoneFunc on_done)>;
+
+// Defines a function prototype to make an asynchronous Report call to
+// the mixer server.
+using TransportReportFunc = std::function<void(
+    const ::istio::mixer::v1::ReportRequest& request,
+    ::istio::mixer::v1::ReportResponse* response, DoneFunc on_done)>;
+
+// Defines a function prototype to make an asynchronous Quota call to
+// the mixer server.
+using TransportQuotaFunc = std::function<void(
+    const ::istio::mixer::v1::QuotaRequest& request,
+    ::istio::mixer::v1::QuotaResponse* response, DoneFunc on_done)>;
+
+// Defines the options to create an instance of MixerClient interface.
+struct MixerClientOptions {
+  // Default constructor with default values.
+  MixerClientOptions() {}
+
+  // Constructor with specified option values.
+  MixerClientOptions(const CheckOptions& check_options,
+                     const ReportOptions& report_options,
+                     const QuotaOptions& quota_options)
+      : check_options(check_options),
+        report_options(report_options),
+        quota_options(quota_options) {}
+
+  // Check options.
+  CheckOptions check_options;
+
+  // Report options.
+  ReportOptions report_options;
+
+  // Quota options.
+  QuotaOptions quota_options;
+
+  // Transport functions are used to send request to mixer server.
+  // It can be implemented many ways based on the environments.
+  // If not provided, the GRPC transport will be used.
+  TransportCheckFunc check_transport;
+  TransportReportFunc report_transport;
+  TransportQuotaFunc quota_transport;
+};
+
+class MixerClient {
+ public:
+  // Destructor
+  virtual ~MixerClient() {}
+
+  // The async call.
+  // on_check_done is called with the check status after cached
+  // check_response is returned in case of cache hit, otherwise called after
+  // check_response is returned from the Controller service.
+  //
+  // check_response must be alive until on_check_done is called.
+  virtual void Check(const ::istio::mixer::v1::CheckRequest& check_request,
```

Comment:
Transport layer has timeout which can serve as cancellation. ESF and ESP have been using this kind of API.  So far we are OK.